### ISSUE TITLE
Load snapshot of Topic back in time

### ISSUE THREAD

**Issue by keyserj** [amelioro/ameliorate#371]
**Is your feature request related to a problem? Please describe.**
I frequently want to make large changes to a diagram without while preserving the ability to see what the previous structure was. Right now, in order to do that, I need to clone the diagram, and add some suffix like "[topic-name]-refactor".

One consistent example of this is after a discourse session - wanting to preserve what we had at the end vs refactor it afterwards.
Another example is when showing the difference between two ways of organizing the same information - like [entirely via a claim tree](https://ameliorate.app/keyserj/debate-marxist-revolution-tree) vs [using structure nodes](https://ameliorate.app/keyserj/debate-marxist-revolution-diagram).

**Describe the solution you'd like**
Add "history" section in Topic details, perhaps with a "Create snapshot now" button, and ability to load/delete/rename snapshots. Loading a snapshot would get all the nodes and edges from that specific point in time and populate the diagram with those - to be interacted-with as if it were the current diagram (but without editing - if you want to edit, download the diagram and re-upload).

Wouldn't need history per-node at this point, but that would probably also be useful (at least to see something like "text changed x->y on DD/MM/YYYY").

**Questions to answer**
- Could make a separate `[Model]Versions` table to make current reads easy?
- Which models to version?
  - Yes: Node, Edge
  - Maybe?: Score
  - No: Comments
- Should snapshots be taken manually, or also automatically? Could do 1 every X (100?) changes

**Describe alternatives you've considered**

**Additional context**

**Technical ideas**
Probably do some googling to look into how others have implemented this type of thing.

One idea:
- Add "created_at" and only create new records, never update
- All reads need to group by UUID and get the latest one
- Topic Snapshot just consists of a name, a topic id, and a time
- Loading a snapshot would call GetData endpoint with a time, which pulls all records based on created_at <= time

**Comment by keyserj** [comment]
Adding topic history and loading a snapshot seem like they should probably be designed together, but these can be split into two tickets if that seems better later.

**Comment by prototyperspective** [comment]
Important issue. Again, a wiki-approach would be really useful for many reasons such as transparency, familiarity, potential interoperability/embeddability, enabling related features, ability to use/adapt some of many existing wiki gadgets/scripts, good revert functionality, etc. I don't think a 'snapshot' approach is a good idea. Why not just have all the changes logged and revertable? Then one can pick any version not just a specific one where somebody created a snapshot and these can be distinguished easily via the timestamps (e.g. latest version before the discourse session). One can also see the auto-generated edit summary to differentiate between changes and pick the right version or even go through the diffs one at a time with the next diff button. If you think a snapshot-type feature would be needed what about adding the ability to tag revisions – e.g. version 0.2 to some revision after large refactoring was finished.

I don't know and haven't yet looked into how the data is stored – if it's json one could show simply diffs of the json file as if it's a normal wiki page (these diffs show only the changed parts not the whole doc). At a later point one could convert or parse it into a more readable format. For example [here](https://argdown.org/) at the map click on source. Each node has its own history but all are part of the same json I think so the revision history for the whole problem tree would show the diffs of the entire json file. A more readable format would be e.g. replacing the diff of some changed json for a newly added connection line with "added link of type {type} between node {title of node} and node {title of node}"...or replacing the diff of an edited description of one node with just the diff of the description field value instead of the json.

This is a key component so yes it could be a bit tricky but lots of other things depend on this feature such as the ability to watch changes to bookmarked problem trees, edit reversion, user contributions pages, content attribution, functionality to list unseen changes. Below is how the unseen changes are listed on a debate site
on Kialo, a per problem tree and site-wide feed could be based on these histories. When clicking on any of these cards in the sidepanel it opens the associated node in the map and marks it as seen ![k](https://github.com/user-attachments/assets/36b2b2a4-2ae9-4687-afcb-60fe9a8890ea) On Wikipedia, there is a separate Watchpage (not embedded via a sidepanel) and when going from there to the history page of an article there is a green dot next to all unseen changes to that from there one can check the changes. I think it would be best to have a History of all revisions of the problem tree (later one could do more with that) and show the unseen diffs somewhere on the problem map page. I think it makes most sense for people to go through all changes on a per node-level instead of seeing a diff of all changes at once (the map could however color all of these in some way) – the changes could be grouped by node so if a node (id 123) was added and then edited, there would only be one item in the problem map History for node 123 for users to check – nevertheless all changes would be in the Page revision history ordered chronologically but that is mostly used for reverting specific changes.

**Comment by keyserj** [comment]
> I don't think a 'snapshot' approach is a good idea. Why not just have all the changes logged and revertable?

The intent behind a snapshot is that notable moments in the topic's history are marked, for the sake of viewing and understanding the evolution of this topic. Only tracking all changes does not enable this. In the `Technical ideas` section, I mention that a snapshot is essentially a label and a time.

> what about adding the ability to tag revisions

I think tags are a good idea that would equally solve the goal of snapshots, assuming it'd be easy to view a list of tags without having to scroll through a history of potentially 100s/1000s of changes. A nice benefit of tags is that they're a common solution to come with history tracking mechanisms, so potentially we could just get tags for free if we're able to implement history with an existing tech.

> haven't yet looked into how the data is stored

In the backend it's stored via relational tables ([schema file](https://github.com/amelioro/ameliorate/blob/7769de5bafed4f1ed1736e8b276f285c69922243/src/db/schema.prisma), [schema diagram](https://github.com/amelioro/ameliorate/blob/7769de5bafed4f1ed1736e8b276f285c69922243/design-docs/database-schema.md)) - when it's loaded in the frontend, it's loaded across a few stores which are each JSON ([here's the store for nodes/edges/scores](https://github.com/amelioro/ameliorate/blob/7769de5bafed4f1ed1736e8b276f285c69922243/src/web/topic/store/store.ts#L55-L60)).

> show simply diffs of the json file

Showing text diffs seems like a great shorter-term solution - I was envisioning UI diffs, like showing a diagram with colored red/green deletions/additions, which would likely be a ton more work.

> replacing the diff of some changed json for a newly added connection line with [...]

This would be nice to skim through a bunch of commits, kind of like generated commit messages. I guess diagram diffs seem particularly useful when viewing more than one change at a time.

> lots of other things depend on this feature

Yeah

> a per problem tree and site-wide feed could be based on these histories. When clicking on any of these cards in the sidepanel it opens the associated node in the map and marks it as seen.

> have a History of all revisions of the problem tree (later one could do more with that) and show the unseen diffs somewhere on the problem map page

Good ideas

> most sense for people to go through all changes on a per node-level instead of seeing a diff of all changes at once

I think having node-level changes is important, but also map-level seem good too, at least when viewed with a diagram diff to easily see added/removed things

> grouped by node so if a node (id 123) was added and then edited, there would only be one item in the problem map History for node 123 for users to check

This seems maybe like sugar on top, I'm not actually sure if this is necessary

**Comment by prototyperspective** [comment]
> The intent behind a snapshot is that notable moments in the topic's history are marked, for the sake of viewing and understanding the evolution of this topic.

I don't see why this would ever be useful while I do think saving and being able to see the entire history instead of just snapshots is needed. Snapshots work for some things but for a continuously developed problem map I think it makes little sense – people would just pick a random point in time or version and make a snapshot then. Even for problem maps that aren't continuously developed, e.g. private maps about some internal matters created by a small-team like a company, one can simply see the entire history instead of some snapshots so I don't see what the use of it would be and don't know if you think of this as something in addition to having the full revision history or a replacement for that (I thought the latter).

> I think tags are a good idea that would equally solve the goal of snapshots, assuming it'd be easy to view a list of tags without having to scroll through a history of potentially 100s/1000s of changes

I think snapshots would not help with that. One could differentiate between changes that are additions of new nodes and other edit types. In any case it seems like the Full revision history is only looked at very rarely since people would rather use the changes feed where changes could be merged by node (one item for eg 4 text edits 1 node type change and the addition of the node instead of 6 diffs to look at). MediaWiki [has tags](https://en.wikipedia.org/wiki/Special:Tags) but it's for changes probably not lightweight enough to be used. I will look for lightweight packages that could enable revision history features later.

>I was envisioning UI diffs, like showing a diagram with colored red/green deletions/additions
>I guess diagram diffs seem particularly useful when viewing more than one change at a time.

Would recommend to not spend time developing an UI for a full map diff view, I think people would go through diffs one at a time and any display of diffs in the full map view wouldn't really be used. For simple text diffs there probably already are many packages that provide diff views for these. For viewing multiple changes in combination, people can see a node diff than navigate to one of its few attached nodes that was also added/changed and then navigate back and so on (multiple one at a time instead of reviewing multiple at once within one view).

>also map-level seem good too, at least when viewed with a diagram diff to easily see added/removed things

I think it would be too much on one page and only people who are very familiar with the site for long times ie powerusers would make frequent use of the full-map instead of some simplified view, especially when it's made even more complex and information-rich than the full map already is via some diff view. Two different node highlightings for either unseen changed nodes and unseen newly added nodes would make sense – eg red background/border around added nodes/connections and blue border glow around changed nodes.

>This seems maybe like sugar on top, I'm not actually sure if this is necessary

Otherwise it would be difficult to understand changes and having to go through many changes. If an item was added then edited then moved, then edited 4 times, then removed, then readded and moved to another location, and then edited 5 times (with one because of forgetting to add a point at the end) do you think it would really be better for the user to go through each of the 15 changes one after another? This is even more problematic if they can't see there's subsequent changes and think the node needs to be moved after seeing just the first or the first few changes. It's merged by just diffing with the latest version. Maybe the description was a bit unclear – the full map history (Page revision history) would show the normal full history, this was only referring to the feed that people go through to review the changes. Could be you're right that it doesn't need to combined but I do think that somehow changes need to be merged. In the screenshot above when one clicks on one of the cards in the unseen changes feed one goes to changed node and sees the node History of all changes to it, then all of the associated cards in the unseen changes feed are marked as seen (here it's two). I think when nodes are edited there multiple times in a row it shows only as one diff but maybe that's only if it's edited before being approved (added in).
![diff2](https://github.com/user-attachments/assets/d195727a-5e96-4185-b80e-ed60b7c77318)

**Comment by keyserj** [comment]
> being able to see the entire history instead of just snapshots is needed

> don't know if you think of this as something in addition to having the full revision history or a replacement

My intent behind snapshots would _not_ replace having the entire history, they would be _in addition to_ a comprehensive history. I agree with the needs for a comprehensive history for sure, but, having made major changes to a Topic after specific events, I also see value in snapshots. It also seems like snapshots would have a somewhat low implementation cost.

After re-reading this issue description, I see how this confusion could've happened. I think I was assuming that history was a prerequisite to snapshots, but didn't document that. I've edited this issue to be about History, and created #563 for the snapshot aspect.

> MediaWiki has tags but it's for changes probably not lightweight enough to be used

Good to know

> I think people would go through diffs one at a time

> For viewing multiple changes in combination, people can see a node diff than navigate to one of its few attached nodes that was also added/changed and then navigate back

I'm having a hard time envisioning this being better if there's 20+ changes since you've viewed last? Maybe this would be uncommon if a Topic is relatively solid, but it doesn't seem like it'd be uncommon if a Topic is still relatively new. I think a visual diff would make it easier to encapsulate all of that at once - though a text diff seems good enough for an initial implementation.

One note: even for a text diff, some changes might be hard to follow. When I've looked at a Topic's JSON, nodes are relatively easy to follow, but edges aren't so easy - they just have a node's id for source/target, which doesn't say much without going to that node. It might be good to come up with an easier-to-understand text representation, maybe something like Kialo's exported txt would be possible and diffable https://support.kialo.com/en/hc/exporting-a-discussion/. This might also be relevant #383.

> do you think it would really be better for the user to go through each of the 15 changes one after another?

This is where I was envisioning it'd be ideal to just see one diff of all the changes since last visiting, because that would inherently jump past all the intermediate changes.

> It's merged by just diffing with the latest version

> full map history (Page revision history) would show the normal full history, this was only referring to the feed

I think we're envisioning similar things. 1. be able to see all individual changes, without merging, and 2. be able to see a diff between any two points in time, with specifically some easily-accessed "since last viewed" view.

I wasn't thinking much about a feed, but I see value in having it.

> In the screenshot above [...]

I see. This UX seems reasonable to me

**Comment by prototyperspective** [comment]
Ah okay understood now,
>I'm having a hard time envisioning this being better if there's 20+ changes since you've viewed last?

I meant people would look at changes of one node at a time either like in the image I attached where the full node revision history can be seen and/or a since-last-seen diff for the node. I don't think they would look at the map to see all of the changes at once or at least it's not needed much if the former method exists. Of course it one goes through one's unseen changes feed it should start with the nodes closest to seen / already existing nodes instead of showing nodes added far down a chain of nodes that you haven't yet seen were newly added. This way you just go through the nodes and also one could see the nodes directly beneath a node one is currently look at so if one or multiple of these were changes one could still view that in the one node at a time viewer. I guess it depends on which UI view mode people would use. Not sure what you mean with visual diffs as visual diffs would also be in one node at a time diff-viewing and look roughly like those in Wikipedia.

>edges aren't so easy - they just have a node's id for source/target

Good point/caveat – one could use some sort of API to load node id1 and node id2 in the diff view so one can see both nodes and compare, maybe with a button "go to / show node in map" for each.

>it'd be ideal to just see one diff of all the changes since last visiting

This works for texts but not for large maps of interconnected nodes or at least not with 5+ changes, especially not 300 changes.

>I think we're envisioning similar things. 1. be able to see all individual changes, without merging, and 2. be able to see a diff between any two points in time, with specifically some easily-accessed "since last viewed" view.

Yes but to add: 1 is about the entire map and individual nodes 2 is I think only useful for individual nodes where for the full map I think (seems like you'd find such a full map diff useful somehow) it would only be useful to highlight which nodes have unseen changes in the full map but not much for some kind of diff of all the changes. Maybe a full diff for full map like a change report would be the best quickest way to see the changes but I think it would be quite difficult to develop as each diff somehow needs lots of context for understanding the change etc. One can't just show (This node has text changed from x to y; This node was added; etc as the context like the connected nodes and the connections matter a lot.)

**Comment by keyserj** [comment]
> Not sure what you mean with visual diffs as visual diffs would also be in one node at a time diff-viewing and look roughly like those in Wikipedia.

What I mean by "visual diff" is something like this https://youtu.be/ZHkofuoOxHQ?t=284 - though the toggle button doesn't seem necessary and I'd expect "Sheep Incubators" maybe to be orange or showing a text diff like with code diffing, since only that node was changed, not created. When I say "visual" I mean not just text, but like I imagine these nodes being on a canvas that's interactive, you can click on nodes to then see individual node history (like a text-list of changes for that one node) etc.

> one could use some sort of API to load node id1 (node previously linked to) and node id2 (node now linked to) in the diff view so one can see both nodes and compare, maybe with a button "go to / show node in map" for each.

Oh yeah that seems like a good idea.

> 1 is about the entire map and individual nodes

That's what I'm thinking too

> 2 is I think only useful for individual nodes where for the full map I think (seems like you'd find such a full map diff useful somehow) it would only be useful to highlight which nodes have unseen changes in the full map but not much for some kind of diff of all the changes

> One can't just show a report like This node has text changed from x to y; This node was added; etc… as the context like the connected nodes and the connections matter a lot.

Hopefully the above youtube clip and my description helps clarify this. I can see this being confusing up to this point, and agree that a text-diff of the whole map is probably not as useful.
